extension/src/plots/model/collect.ts

```diff
@@ -376,6 +466,93 @@ export const collectSelectedComparisonPlots = ({
   return acc
 }
 
+const getSelectedImgComparisonPlotClasses = ({
+  comparisonClassesSelected,
+  img,
+  path
+}: {
+  comparisonClassesSelected: ComparisonClassesSelected
+  img: ImagePlot
+  path: string
+}) => {
+  const imgClasses: ComparisonPlotClass[] = []
+
+  for (const [label, boxes] of Object.entries(img.boundingBoxes || {})) {
+    const selectedState = comparisonClassesSelected[path]?.[label]
+
+    if (selectedState === false) {
+      continue
+    }
+
+    imgClasses.push({
+      boxes,
+      label
+    })
+  }
+
+  return imgClasses
+}
+
+const collectedSelectedPathComparisonPlotClasses = ({
+  acc,
+  comparisonClassesSelected,
+  id,
+  comparisonData,
+  path
+}: {
+  comparisonClassesSelected: ComparisonClassesSelected
+  acc: ComparisonPlotClasses
+  comparisonData: ComparisonData
+  path: string
+  id: string
+}) => {
+  for (const img of comparisonData[id][path]) {
+    const imgClasses = getSelectedImgComparisonPlotClasses({
+      comparisonClassesSelected,
+      img,
+      path
+    })
+
+    if (imgClasses.length === 0) {
+      return
+    }
+
+    if (!acc[id]) {
+      acc[id] = {}
+    }
+
+    acc[id][path] = imgClasses
+  }
+}
+
+export const collectSelectedComparisonPlotClasses = ({
+  comparisonClassesSelected,
+  comparisonData,
+  paths,
+  selectedRevisionIds
+}: {
+  comparisonClassesSelected: ComparisonClassesSelected
+  comparisonData: ComparisonData
+  paths: string[]
+  selectedRevisionIds: string[]
+}) => {
+  const acc: ComparisonPlotClasses = {}
+
+  for (const id of selectedRevisionIds) {
+    for (const path of paths) {
```

Comment:
What I meant by removing nested loops was not moving one inside of another function (I know realize we have 4 nested loops total), but can we change the code to drop some or at least make them more performant. 

I see that in the fourth nested loop, we check for selected state. Maybe for a start, we could loop only through `paths` that are also in `comparisonClassesSelected`, since we only care about  labels in `comparisonClassesSelected[path]`.

There might be other optimizations to be made as well.

Reply:
> What I meant by removing nested loops was not moving one inside of another function (I know realize we have 4 nested loops total), but can we change the code to drop some or at least make them more performant.

Thanks for the clarification! 

> I see that in the fourth nested loop, we check for selected state. Maybe for a start, we could loop only through paths that are also in comparisonClassesSelected, since we only care about labels in comparisonClassesSelected[path].

`comparisonClassesSelected` only holds classes that the user has toggled directly, it doesn't hold every visible class. Unless I'm missing something, we can't rely on it to detect which paths need to be looped over to collect classes.

The four loops are:
1. looping over each revision (up to 7)
2. looping over each path (could be a lot depending on how many plots the user has)
3. looping over each path img (will usually be just one, unless it's an image step plot)
4. looping over each class, gathering only the selected class boxes

I don't see a way to get rid of the first three loops, since for us to gather a plot's classes, we need to check for classes in each single image and collect the data. Atleast the third loop will likely just be a single item array. 

We could get rid of the fourth loop by passing all image box data to the frontend and have `ComparisonTableBoundingBoxImg` filter out the boxes, but that means passing larger arrays that the component needs to loop over. Seems like we would want to limit the data passed to the frontend as much as possible but please correct me if I'm wrong.

Another alternative is having `collectSelectedComparisonPlots` gather class information along with the selected comparison plot data. I chose not to do this since that would make the function more complex, but if we think it's worth it for performance, I'm alright with doing that. 

### TL;DR

I'm not seeing a good way to get rid of any of our four loops. I'll review the loops and see what we can do to optimize them for now. 

@sroy3, what do you think?

Reply:
Reviewed optimizing the loops! Instead of looping over every revision/path in `comparisonData`, we could start by looping over the already formed `plots` array. The `plots` already have a collected `classDetails` that tells you if a path has selected classes or not, lettings us skip early. Aka:

1. looping over each `plots` path (skip over if `classDetails` is empty)
2. looping over each revision
3. looping over each path img (will usually be just one, unless it's an image step plot)
4. looping over each class, gathering only the selected class boxes

Not seeing any other optimizations besides the two ideas I mentioned in my previous comment (have `collectSelectedComparisonPlots` collect box data, have `ComparisonTableBoundingBoxImg` do the box filtering).

Reply:
One thing I'm seeing is this:
```
if (selectedState === false) {
  continue
}
```
```
if (imgClasses.length === 0) {
  return
}
```
    
Why do we have to go through four loops to return nothing when we know it before hand? I would make this the first loop (the selected classes). 

If there are no other possible optimizations, then I guess it's fine to keep as is.

Reply:
> if (selectedState === false) {
  continue
}
> Why do we have to go through four loops to return nothing when we know it before hand? I would make this the first loop (the selected classes).

Aren't we already getting the selected classes [in the first loop](https://github.com/iterative/vscode-dvc/pull/5250/files#diff-b8591ca0f7ff9633208614960ef48ca0fd0bcc1fb055531bbf943a8250a14d86R519)? In fact the line `if (selectedState === false) {...` no longer exists in the code. Did Github direct you the latest changes or is there something I'm missing here? 

> One thing I'm seeing is this:
> if (imgClasses.length === 0) {
  return
}

Good catch on `if (imgClasses.length === 0) {...`! I didn't spot the fact that we might be needlessy looping over selected classes if an img has no attached boxes to begin with. I'll update the code to check for `img.boxes` before running `getSelectedImgComparisonPlotClasses`.